CONTRIBUTING.md

```diff
@@ -7,100 +7,198 @@
+the following tags prepended to the title:
+
+- `[WIP]` - Work In Progress. The PR will change, so while there is no
+  immediate need for review, although the submitter still might appreciate it.
+- `[RFC]` - Request For Comment. The PR needs reviewing and/or comments.
+- `[RDY]` - The PR is ready to be merged. The PR has been reviewed by at least
+  one other person and has no outstanding issues.
 
-* `[WIP]` - Work In Progress. The pull request will change, and there is no need
-  to review it yet.
-* `[RFC]` - Request For Comment. The request needs reviewing and/or comments.
-* `[RDY]` - The request is ready to be merged. The request must have been
-  reviewed by at least one person and have no outstanding issues.
-* Default label is assumed to be `[WIP]` if there's no indication otherwise.
+Assuming the above criteria has been met, feel free to change your PR's tag
+yourself, as opposed to waiting for a contributor to do it for you.
 
 #### Branching & history
 
-- Use a feature branch, not master.
-- Rebase your feature branch onto (upstream) master before raising the PR.
-- Keep up to date with changes in (upstream) master so your PR is easy to merge.
-- Try to actively tidy your history: combine related commits with interactive
-  rebasing etc. If your PR is still `[WIP]` don't be afraid to force-push to
-  your feature branch to tidy your history.
+- Do ***not*** work on your PR on the master branch, [use a feature branch
+  instead][feature-branch]. ***TODO:*** this feature branch article is really
+  long, a shorter one would be nicer.
+- [Rebase your feature branch onto][git-rebasing] (upstream) master before
+  opening the PR.
+- Keep up to date with changes in (upstream) master so your PR is easy to
+  merge.
+- [Try to actively tidy your history][git-rewriting-history]: combine related
+  commits with interactive rebasing, split monolithic commits into multiple
+  ones, etc. If your PR is still `[WIP]` don't be afraid to force-push to your
+  feature branch to tidy your history.
 
 ### For code PRs
 
 #### Testing
 
-- We are unlikely to merge your PR if the Travis build fails.
-- The Travis build does not currently run the tests under valgrind, but you
-  are encouraged to do so locally.
+We are unlikely to merge your PR if the Travis build fails:
+
+- Travis builds are compiled with the [`-Werror`][gcc-warnings] flag, so
+  if your PR introduces any compiler warnings then the Travis build will fail.
+- If any unit tests fail, the Travis build will fail. They can be run
+  locally with `make test`, which is generally much faster than waiting on
+  Travis, so should be done prior to submitting your PR. Unit tests
+  passing locally doesn't guarantee they'll pass in your Travis build, as
+  multiple different builds using different platforms and compilers are
+  used to test your code.
+- The Travis build does not currently run the tests under
+  [valgrind][valgrind], but you are encouraged to do so locally.
```

Comment:
There's `USE_VALGRIND`, but it's undocumented. I guess I'll add it here.

@tarruda just to verify, does it only run under `oldtests`?